Reoganize and partially rewrite the main normative sections (#9)
ietf-wg-spice/draft-ietf-spice-sd-cwt: draft-ietf-spice-sd-cwt.md

: A CWT with claims enabling selective disclosure with key binding.

Selective Disclosure Key Binding Token (SD-CWT-KBT):
: A CWT used to demonstrate possession of a confirmation method, associated to an SD-CWT.

Issuer:
: An entity that produces a Selective Disclosure CBOR Web Token.

Holder:
: An entity that presents a Selective Disclosure CBOR Web Token which includes a Selective Disclosure Key Binding Token.

Verifier:
: An entity that validates a Partial or Full Disclosure by a holder.

Partial Disclosure:
: When a subset of the original claims protected by the Issuer, are disclosed by the Holder.

Full Disclosure:
: When the full set of claims protected by the Issuer, is disclosed by the Holder.

Salted Disclosed Claim:
: A salted claim disclosed in the unprotected header of an SD-CWT.

Digested Salted Disclosed Claim / Blinded Claim Hash:
: A hash digest of a Salted Disclosed Claim.

Blinded Claim:
: Any Redacted key or Redacted element which has been replaced in the
CWT payload by a Blinded Claim Hash.

Redacted key:
: The hash of a claim redacted from a map data structure.

Redacted element:
: The hash of an element redacted from an array data structure.

> Presented Disclosed Claimset:

Reviewer: ```suggestion
Presented Disclosed Claims Set:
```

: The CBOR map containing zero or more Redacted keys or Redacted elements.

> Validated Disclosed Claimset:

Reviewer: ```suggestion
Validated Disclosed Claims Set:
```

: The CBOR map containing all mandatory to disclose claims signed by the issuer, all selectively disclosed claims presented by the holder, and ommiting all instances of redacted_keys and redacted_element claims that are present in the original sd_cwt.

# SD-CWT definition

SD-CWT is modeled after SD-JWT, with adjustments to align with conventions

> in CBOR and COSE. An SD-CWT MUST include a `typ` claim of
>  "application/sd-cwt" in the protected header of the SD-CWT.

Reviewer: ```suggestion
in CBOR and COSE. An SD-CWT MUST include the protected header parameter
`typ` {RFC9596} with the value `application/sd-cwt` in the SD-CWT.
```

Author: > 

```suggestion
in CBOR and COSE. An SD-CWT MUST include the protected header parameter
`typ` with the value "application/sd-cwt" in the SD-CWT.
```

CWT (the disclosures).

Any party with a Salted Disclosed Claim can generate its hash, find that
hash in the CWT payload, and unblind the content. However a Verifier with the hash cannot reconstruct the corresponding blinded claim without
disclosure of the Salted Disclosed Claim.

## Types of blinded claims

Salted Disclosed Claims for named claims are structured as a 128-bit salt, the name of the redacted element, and the disclosed value. For Salted
Disclosed Claims of items in an array, the name is omitted.

~~~ cddl
salted = salted-claim / salted-element / decoy
salted-claim = [
  bstr .size 16,     ; 128-bit salt
  (int / text),      ; claim name
  any                ; claim value
]
salted-element = [
  bstr .size 16,     ; 128-bit salt
  any                ; claim value
]
decoy = [
  bstr .size 16,     ; 128-bit salt
]

; a collection of Salted Disclosed Claims
salted-array = [ +bstr .cbor salted ]
~~~

When a blinded claim is a key in a map, its blinded claim hash is added to a `redacted_values` array claim in the CWT payload that is at the same level of hierarchy as the key being blinded.

When blinding an individual item in an array, the value of the item is

> replaced with a dict containing only the special key "...".

Reviewer: "dict" isn't a CBOR or COSE term.

Author: fixed

> replaced with a dict containing only the special key "...".

Author: ```suggestion
replaced with a CBOR map containing only the special key "...".
```

~~~ cddl
redacted_element = { "...": any }
~~~

Blinded claims can be nested. For example, both individual keys in the
`address` claim, and the entire `address` element can be separately blinded.
An example nested claim is shown in **TODO iref**.

Finally, an issuer may create "decoy digests" which look like a blinded
claim hash but have only a salt. Decoy digests are discussed in **TODO iref**.

# SD-CWT Issuance

How the Holder communicates to the Issuer to request a CWT or an SD-CWT is
out-of-scope of this specification. Likewise, how the Holder determines
which claims to blind or to always disclose is a policy matter which is
not discussed in this specification. This specification defines the format
of an SD-CWT communicated between an Issuer and a Holder in this section, and describes the format of an SD-CWT communicated between a Holder and a
Verifier in the next section.

The unprotected header MUST contain an `sd_claims` section with a Salted
Disclosed Claim for *every* blinded claim hash present anywhere in the

> payload. The payload MUST also include a key confirmation element (`cnf`)

Reviewer: ```suggestion
payload. The payload MUST also include a key confirmation element (`cnf`) {RFC8747}
```

Author: > 

```suggestion
payload. The payload MUST also include a key confirmation element (`cnf`) {{!RFC8747}}
```

for the Holder's public key, and an `sd_alg` claim identifying the
algorithm used to hash the Salted Disclosed Claims.

## Issuer generation

The Issuer follows all the requirements of generating a valid CWT.

> The Issuer MUST implement COSESign1 using an appropriate asymmetric

Reviewer: ```suggestion
The Issuer MUST implement COSE_Sign1 using an appropriate asymmetric
```

signature algorithm / curve combination (for example ES256/P-256 or
EdDSA/Ed25519)

> The Issuer MUST generate a unique cryptographically random salt of at least

Reviewer: ```suggestion
The Issuer MUST generate a unique cryptographically random salt with at least
```

128-bits of entropy for each Salted Disclosed Claim. If the client
communicates a client-generated nonce (`cnonce`) when requesting the SD-CWT,
the Issuer SHOULD include it in the payload.

## Holder validation

Upon receiving an SD-CWT from the Issuer with the Holder as the subject, the
Holder verifies the following:

- the issuer (`iss`) and subject (`sub`) are correct;
- if an audience (`aud`) is present, it is acceptable;

> - the CWT is valid according to the `nbf` and `exp` claims;

Reviewer: ```suggestion
- the CWT is valid according to the `iat` and `exp` claims;
```

Author: I meant `nbf` and `exp`. The current time is greater or equal to `nbf` and not more than `exp`. What comparison would you make with `iat`?

Author: Marking as resolved, to be addressed in Issue #14

   * key => any
}

unprotected-issued = {
   &(sd_claims: TBD1) ^ => bstr .cbor salted-array,
   * key => any
}

sd-payload = {
    ; standard claims
      &(iss: 1) ^ => tstr, ; "https://issuer.example"
      &(sub: 2) ^ => tstr, ; "https://device.example"
    ? &(aud: 3) ^ => tstr, ; "https://verifier.example/app"

>     ? &(exp: 4) ^ => int,  ; 1883000000

Reviewer: Why is "exp" optional?  It seems like it should be required.

Author: RFC8392 is silent about if exp is required in a CWT. It refers to the definition in [Section 4.1.4 of RFC7519](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4) which says `exp` is optional in a JWT.

Reviewer: For what it's worth, *all* claims are optional in JWTs because they are expected to be profiled for particular use cases.  In our case, we may want to define the minimum required claims.

I thought that the "?" syntax above made "exp" optional.  Am I correct?  And similarly, the change above makes "aud" optional.  Is that what we want to do?

Author: I have a legitimate use case for "aud" to be optional (use in MLS), but the "aud" would be required in the Key Binding Token.

For "exp" I cannot think of a use case where we want a never expiring SD-CWT, but I was hesitant to change the optionality and break someone else's use case.

Author: I created Issue #14 for this. I propose we resolve in this PR and incorporate the result of this conclusion in another PR.

    ? &(nbf: 5) ^ => int,  ; 1683000000
      &(iat: 6) ^ => int,  ; 1683000000
      &(cnf: 8) ^ => { * key => any }, ; key confirmation
    ? &(cnonce: 39) ^ => bstr,
    ;
    ; sd-cwt new claims
      # TODO: should sd_hash be included in issued?
      &(sd_hash: TBD3) ^ => bstr,
      &(sd_alg: TBD4) ^ => int,            ; -16 for sha-256

>     ? &(redacted_keys: TBD5) ^ => [ * bstr ],

Reviewer: I would change all instances of "redacted_keys" to "redacted_values".  It is unfortunate that CBOR chose to use the term "key" for the labels of objects, as it inevitably causes readers confusion between cryptographic keys and object keys.

Please let's reserve the term "key" for cryptographic keys.

Author: > I would change all instances of "redacted_keys" to "redacted_values". It is unfortunate that CBOR chose to use the term "key" for the labels of objects, as it inevitably causes readers confusion between cryptographic keys and object keys.
> 
> Please let's reserve the term "key" for cryptographic keys.

"values" doesn't convey the difference between array items map elements. keys isn't really correct either.  How do folks feel about `redacted_elements`?

Author: I proposed `redacted_map_claims` in Issue #15 . Let's please merge this PR while these other issues get discussed.

Author: Marking resolved. Captured in Issue #15

    * key => any
}
~~~

# SD-CWT Presentation

When a Holder presents an SD-CWT to a Verifier, it can disclose
none, some, or all of its blinded claims. If the Holder wishes to disclose any claims, it includes those Salted Disclosed Claims in the `sd_claims` claim of the unprotected header.

~~~ cddl
sd-cwt-presentation = #6.18([
   protected: bstr .cbor sd-protected,
   unprotected-presentation,
   payload: bstr .cbor sd-payload,
   signature: bstr
])

unprotected-presentation = {
   &(sd_kbt: TBD2) ^ => bstr .cbor kbt-cwt,
   ? &(sd_claims: TBD1) ^ => bstr .cbor salted-array,
   * key => any
}
~~~

As described in the CDDL above, a SD-CWT presentation to a Verifier has the

> same syntax as an SD-CWT issued to a Holder, except for its unprotected header.

Reviewer: I would suggest saying what about the unprotected header is different.  Help readers understand without having to look elsewhere in the spec.

Author: I'll take a stab at this tomorrow.

Author: > I'll take a stab at this tomorrow.

How well does the proposed change (starting line 329) address your concern about explaining the unprotected header?

## Creating a Key Binding Token {#kbt}

Regardless if it discloses any claims, the Holder MUST include a Holder key
binding (SD_KBT) {{kbt}} in a `sd_kbt` claim in the unprotected header in
every presentation of an SD-CWT by a Holder to a Verifier.
(The `sd_kbt` claim is absent when the Issuer is providing the

> SD-CWT to the Holder.) An SD-KBT is itself a type of CWT. The `typ` claim of
> "application/sd-kbt" MUST be included in the protected header of the SD-KBT.

Reviewer: ```suggestion
SD-CWT to the Holder.) An SD-KBT is itself a type of CWT. The protected header of the SD-KBT
MUST include the `typ` header parameter with the value `application/sd-kbt`.
```

    ? &(aud: 3) ^ => tstr, ; "https://verifier.example/app"
    ? &(exp: 4) ^ => int,  ; 1883000000
    ? &(nbf: 5) ^ => int,  ; 1683000000
      &(iat: 6) ^ => int,  ; 1683000000
      &(cnf: 8) ^ => { * key => any }, ; key confirmation
    ? &(cnonce: 39) ^ => bstr,
    ;
    ; sd-cwt new claims
      # TODO: should sd_hash be included in issued?
      &(sd_hash: TBD3) ^ => bstr,
      &(sd_alg: TBD4) ^ => int,            ; -16 for sha-256
    ? &(redacted_keys: TBD5) ^ => [ * bstr ],
    * key => any
}
~~~

# SD-CWT Presentation

When a Holder presents an SD-CWT to a Verifier, it can disclose
none, some, or all of its blinded claims. If the Holder wishes to disclose any claims, it includes those Salted Disclosed Claims in the `sd_claims` claim of the unprotected header.

~~~ cddl
sd-cwt-presentation = #6.18([
   protected: bstr .cbor sd-protected,
   unprotected-presentation,
   payload: bstr .cbor sd-payload,
   signature: bstr
])

unprotected-presentation = {
   &(sd_kbt: TBD2) ^ => bstr .cbor kbt-cwt,
   ? &(sd_claims: TBD1) ^ => bstr .cbor salted-array,
   * key => any
}
~~~

As described in the CDDL above, a SD-CWT presentation to a Verifier has the

> same syntax as an SD-CWT issued to a Holder, except for its unprotected header.

Author: ```suggestion
same syntax as an SD-CWT issued to a Holder, except for its unprotected header.
Since the unprotected header is not included in the signature, it will contain
all the Salted Disclosed Claims when sent from the Issuer to the Holder. By
comparison, the unprotected header will include a Key Binding Token and contain
none, some, or all of these Claims when sent from the Holder to the Verifier.
```

Otherwise the SD-CWT is considered valid, and the Validated Disclosed Claimset is now a CWT Claimset with no claims marked for redaction.
Further validation logic can be applied to the Validated Disclosed Claimset, as it might normally be applied to a validated CWT claimset.

# Decoy Digests

**TODO**

# Credential Types

This specification defines the CWT claim vct (for verifiable credential type). The vct value MUST be a case-sensitive StringOrURI (see {{RFC7519}}) value serving as an identifier for the type of the SD-CWT claimset. The vct value MUST be a Collision-Resistant Name as defined in Section 2 of {{RFC7515}}.

This claim is defined COSE based verifiable credentials, similar to the JOSE based verifiable credentials described in Section 3.2.2.1.1 of SD-JWT-VC.

> Profiles built on this specifiation are also encouraged to use more specific media types, as described in {{?I-D.ietf-cose-typ-header-parameter}}.

Reviewer: ```suggestion
Profiles built on this specification are also encouraged to use more specific media types, as described in {{?RFC9596}}.
```

Author: > 

```suggestion
Profiles built on this specification are also encouraged to use more specific media
types, as described in {{!RFC9596}}.
```
